docs/design/features/objc-interop.md

```diff
@@ -0,0 +1,160 @@
+# Objective-C interoperability
+
+This design document describes a singular .NET Platform support scenario for an Objective-C runtime. This design is heavily influenced by existing [Xamarin-macios][xamarin_repo] support which is built on top of Mono's [Embedding API](https://www.mono-project.com/docs/advanced/embedding/).
+
+## Objective-C concepts
+
+There are concepts in Objective-C that require attention for interoperability.
+
+**Message dispatch** <a name="message_disp"></a>
+
+Method dispatch in Objective-C is typically done through message passing. A suite of functions make up the message passing system.
+
+- [`objc_msgSend`](https://developer.apple.com/documentation/objectivec/1456712-objc_msgsend) - Canonical message passing.
+- [`objc_msgsend_fpret`](https://developer.apple.com/documentation/objectivec/1456697-objc_msgsend_fpret) - Message returns a floating point value.
+- [`objc_msgSend_stret`](https://developer.apple.com/documentation/objectivec/1456730-objc_msgsend_stret) - Message returns a structure.
+- [`objc_msgSendSuper`](https://developer.apple.com/documentation/objectivec/1456716-objc_msgsendsuper) - Send message to the super class.
+- [`objc_msgSendSuper_stret`](https://developer.apple.com/documentation/objectivec/1456722-objc_msgsendsuper_stret) - Send message that returns a structure to the super class.
+
+Each of these functions are designed to look up the target method on the object ([`id`](https://developer.apple.com/documentation/objectivec/id)) based on the supplied selector ([`SEL`](https://developer.apple.com/documentation/objectivec/sel)) and then `jmp` to that method ([`IMP`](https://developer.apple.com/documentation/objectivec/objective-c_runtime/imp)). Note that this is a `jmp`, not a `call`. From the perspective of the caller the return is from the target method not the message dispatch function. Looking up the above functions, one will note each are defined with a `void(*)(void)` signature. During the compilation of Objective-C code, the compiler will compute the appropriate function signature, cast the message passing function, and make the call. All Objective-C methods have two implied arguments - the object (i.e. `id`) and the message selector (`SEL`). This means the following Objective-C statement:
```

Comment:
I am not sure this is entirely accurate. It works like this when `OBJC_OLD_DISPATCH_PROTOTYPES` is not defined but that's no longer the default:
https://github.com/dotnet/runtime/issues/48796#issuecomment-786669994

Reply:
@filipnavara This is accurate as far as I can tell. The ABI hasn't changed here, can you elaborate on what you are taking issue with?

Reply:
`Looking up the above functions, one will note each are defined with a `void(*)(void)` signature.`

The [Objective-]C headers now use a signature with vararg parameters (`id _Nullable self, SEL _Nonnull op, ...`) when ` OBJC_OLD_DISPATCH_PROTOTYPES` is not defined. This signature can be called without a cast to more specific delegate. Obviously the underlying implementation in the runtime is still the same. Practically it doesn't make any difference on x64 ABI. It may have an impact on ARM64 ABI where varargs are always passed through stack so it makes me wonder how the Objective-C runtime behaves there.

Reply:
Nevermind. I misread the `#if` in the Apple headers and thought it's the other way. That makes way more sense now in respect to the ARM64 situation.

Reply:
>The [Objective-]C headers now use a signature with vararg parameters (id _Nullable self, SEL _Nonnull op, ...)

Ah. Those aren't varargs in this case - see [`objc_msgSend`](https://developer.apple.com/documentation/objectivec/1456712-objc_msgsend?language=objc). That is a confusing legacy definition of the low level API and was fixed for Objective-C but not for Swift due to some excessiveness concerns.

Reply:
I trust the code more than the documentation when it comes to Apple but in this case I messed up with tracking the conditional logic and flipped it in my head. I only realised the mistake after reading the first sentence of https://www.mikeash.com/pyblog/objc_msgsends-new-prototype.html 😅